reps/2023-03-15-train-api.md

```diff
@@ -0,0 +1,143 @@
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@stephanie-wang
+
+## Design and Architecture
+
+## API Changes
+
+1. Introduce a `resource_limits: ExecutionResources(object_store_memory=2 * GiB)` arg to `ray.air.DatasetConfig`. This effectively enables streaming by default, with a limit of 2GiB.
+
+2. Introduce `Dataset.get_preprocessor_lineage()`, which returns the lineage of preprocessors applied to this Dataset. If multiple preprocessors are applied, it will return a `Chain` of the preprocessors. Non-preprocessor operations on the Dataset are ignored. This function will be used by Train to persist fitted preprocessors with checkpoints.
+
+3. Deprecate the following `DatasetConfig` arguments when streaming is enabled: `max_object_store_memory_fraction`, `global_shuffle`, `fit`, `transform`, `randomize_block_order`, and `per_epoch_preprocessor`, as well as the following `Trainer` arguments: `preprocessor`.
+
+## Example:
+
+### Before
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Create the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+
+# Create the per-epoch preprocessor.
+per_epoch_prep = RandomNoisePreprocessor()
+
+# Trainer applies preprocessing internally via config.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    preprocessor=prep,
+    dataset_config={
+        "train_ds": {
+            "max_object_store_memory_fraction": 0.2,  # Enable streaming.
+            "randomize_block_order": True,
+            "per_epoch_preprocessor": per_epoch_prep,
+        },
+    },
+)
+
+# Checkpoint includes fitted preprocessor.
+best_checkpoint = trainer.fit().checkpoint
+assert best_checkpoint.get_preprocessor() == prep
+```
+
+### After
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Fit the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+prep.fit(base)
+
+# Apply base preprocessing.
+train_ds = base.map_preprocessor(prep)
+train_ds.cache()  # Optional: cache the base data in memory.
+
+# Per-epoch preprocessing.
+train_ds = train_ds \
+    .randomize_block_order() \
+    .map_preprocessor(RandomNoisePreprocessor())
+
+# Trainer doesn't know about preprocessing at all.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    dataset_config={
+        "train_ds": {
+            "resource_limits": ExecutionResources(
```

Comment:
is this resource limit only applicable to `train_ds`? do we need a per-dataset level granularity here, or is a global resource limit enough?

Reply:
Yup. I think per-dataset is probably reasonable, since it would be harder to do global (we don't have that implemented).